klarna-on-demand/src/main/java/com/klarna/ondemand/CryptoImpl.java

```diff
@@ -0,0 +1,116 @@
+    private static final String PREFERENCES_FILE_NAME = "KeyPair";
+    private static final String ALGORITHM = "RSA";
+
+    private static CryptoImpl cyptoInstance;
+
+    private String publicKeyBase64Str;
+    private PublicKey publicKey;
+    private PrivateKey privateKey;
+
+
+    public static Crypto getInstance(android.content.Context context) {
+        if (cyptoInstance == null) {
+            try {
+                cyptoInstance = new CryptoImpl(context);
+            } catch (Exception e) {
+                throw new RuntimeException("Could not initialize " + CryptoImpl.class.getName(), e);
+            }
+        }
+        return cyptoInstance;
+    }
+
+    private CryptoImpl(android.content.Context context) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        SharedPreferences sharedPreferences = context.getSharedPreferences(PREFERENCES_FILE_NAME, Context.MODE_PRIVATE);
+
+        publicKey = readPublicKey(sharedPreferences);
+        privateKey = readPrivateKey(sharedPreferences);
+        if (publicKey == null || privateKey == null) {
+            KeyPair keyPair = generateKeyPair();
+            publicKey = keyPair.getPublic();
+            privateKey = keyPair.getPrivate();
+
+            persistKeyPair(sharedPreferences, keyPair);
+        }
+
+        publicKeyBase64Str = new String(Base64.encode(publicKey.getEncoded(), Base64.DEFAULT));
+    }
+
+    @Override
+    public String getPublicKeyBase64Str() {
+        return publicKeyBase64Str;
+    }
+
+    @Override
+    public String sign(String message) throws NoSuchAlgorithmException, InvalidKeyException, SignatureException {
+        Signature sign = Signature.getInstance("SHA256withRSA");
+        sign.initSign(privateKey);
+        sign.update(message.getBytes());
+        return new String(Base64.encode(sign.sign(), Base64.DEFAULT));
+    }
+
+    private void persistKeyPair(SharedPreferences sharedPreferences, KeyPair keyPair) {
+        SharedPreferences.Editor sharedPreferencesEditor = sharedPreferences.edit();
+
+        persistKey(sharedPreferencesEditor, PUBLIC_KEY, keyPair.getPublic());
+        persistKey(sharedPreferencesEditor, PRIVATE_KEY, keyPair.getPrivate());
+
+        sharedPreferencesEditor.commit();
+    }
+
+    private void persistKey(SharedPreferences.Editor sharedPreferencesEditor, String keyName, Key key) {
+        byte[] keyBytes = key.getEncoded();
+        String keyStr = new String(Base64.encode(keyBytes, Base64.DEFAULT));
+        sharedPreferencesEditor.putString(keyName, keyStr);
+    }
+
+    private KeyPair generateKeyPair() throws NoSuchAlgorithmException {
+        KeyPairGenerator kpg = KeyPairGenerator.getInstance(ALGORITHM);
+        kpg.initialize(512);
+        return kpg.genKeyPair();
+    }
+
+    private PublicKey readPublicKey(SharedPreferences sharedPerfernces) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        String pubKeyStr = sharedPerfernces.getString(PUBLIC_KEY, null);
```

Comment:
maybe use the same behaviour as in iOS: create tag variable: `private static final KEY_TAG = "klarna";` and get package name from context: `context.getPackageName()` ?